src/network/networkd-address.c

```diff
@@ -20,6 +21,88 @@
 
+#define ULA_ADDRESS_APP_ID SD_ID128_MAKE(13,ac,83,a7,d6,3f,49,79,94,79,5d,0c,21,3b,bd,7d)
+
+/* Unique Local IPv6 Unicast Addresses https://tools.ietf.org/rfc/rfc4193.txt */
+int generate_ipv6_ula_address(Link *link, uint16_t subnet_id, Address **ret) {
+        _cleanup_(address_freep) Address *address = NULL;
+        _cleanup_(khash_unrefp) khash *h = NULL;
+        uint8_t key[16], eui64[8], global_id[5];
+        const void *p;
+        uint64_t ntp;
+        int r;
+
+        r = address_new(&address);
+        if (r < 0)
+                return log_link_error_errno(link, r, "Could not allocate address: %m");
+
+        address->family = AF_INET6;
+
+        /* 1. Obtain the current time of day in 64-bit NTP format [NTP]. */
+        ntp = (int64_t) now(CLOCK_MONOTONIC);
+        memcpy((uint8_t *) &key, &ntp, sizeof(uint64_t));
+
+        /* 2) Get EUI-64 from the interface specified */
+        eui64[0]  = link->mac.ether_addr_octet[0];
+        eui64[0] ^= 1 << 1;
+        eui64[1]  = link->mac.ether_addr_octet[1];
+        eui64[2] = link->mac.ether_addr_octet[2];
+        eui64[3] = 0xff;
+        eui64[4] = 0xfe;
+        eui64[5] = link->mac.ether_addr_octet[3];
+        eui64[6] = link->mac.ether_addr_octet[4];
+        eui64[7] = link->mac.ether_addr_octet[5];
+
+        /*  3) Concatenate the time of day with the system-specific identifier
+         *   in order to create a key.
+         */
+        memcpy((uint8_t *) &key + sizeof(uint64_t), (void *) &eui64, sizeof(eui64));
+
+        /* 4) Compute an SHA-1 digest on the key as specified in [FIPS, SHA1];
+         *  the resulting value is 160 bits.
+         */
+
+        r = khash_new(&h, "sha1");
+        if (r < 0)
+                return r;
+
+        r = khash_put(h, &key, sizeof(key));
+        if (r < 0)
+                return r;
+
+        r = khash_digest_data(h, &p);
+        if (r < 0)
+                return r;
+
+        /* 5) Use the least significant 40 bits as the Global ID. Not exactly just chop off. */
+        memcpy(&global_id, p, MIN(khash_get_size(h), sizeof(global_id)));
+
+        /* 6) Concatenate FC00::/7, the L bit set to 1, and the 40-bit Global
+         *    ID to create a Local IPv6 address prefix.
+         */
+
+        r = in_addr_prefix_from_string("fc00::/7", AF_INET6, &address->in_addr, &address->prefixlen);
+        if (r < 0)
+                return r;
+
+        /* 3.1.  Format
+         *      The Local IPv6 addresses are created using a pseudo-randomly
+         *       allocated global ID.  They have the following format:
+         *
+         *       | 7 bits |1|  40 bits   |  16 bits  |          64 bits           |
+         *       +--------+-+------------+-----------+----------------------------+
+         *       | Prefix |L| Global ID  | Subnet ID |        Interface ID        |
+         *       +--------+-+------------+-----------+----------------------------+
+         */
+        address->in_addr.in6.s6_addr[0] |= 0x01;
```

Comment:
Why set this bit to 1?

Reply:
section 3.2.2. 
  6) Concatenate FC00::/7, the L bit set to 1, and the 40-bit Global
    D to create a Local IPv6 address prefix.

Reply:
Might want to indicate that in the comment.

Reply:
That's in no 6. Probably you missed.

Reply:
In my mind, the comment for 6) was only to be applied to that particular line of code.

With that in mind, maybe indicate that the L bit set to 1 is done in the next chunk of code down, not in the immediate code chunk.